DockerSettings.php

```diff
@@ -281,11 +297,17 @@
 $dockerLoadExtensions = getenv( 'MW_LOAD_EXTENSIONS' );
 if ( $dockerLoadExtensions ) {
 	$dockerLoadExtensions = explode( ',', $dockerLoadExtensions );
+	$dockerLoadExtensionsNoRegistration = array_intersect( DOCKER_EXTENSIONS_NO_REGISTRATION, $dockerLoadExtensions );
 	$dockerLoadExtensions = array_intersect( DOCKER_EXTENSIONS, $dockerLoadExtensions );
 	if ( $dockerLoadExtensions ) {
 		wfLoadExtensions( $dockerLoadExtensions );
 		$dockerLoadExtensions = array_combine( $dockerLoadExtensions, $dockerLoadExtensions );
 	}
+	if ( $dockerLoadExtensionsNoRegistration ) {
+		foreach ( $dockerLoadExtensionsNoRegistration as $extension ) {
+			require_once "$wgExtensionDirectory/$extension/$extension.php";
+		}
+	}
```

Comment:
Until there is a reason I believe this could be solved in a bit different way keeping, the `DOCKER_EXTENSIONS` in place and avoiding the need of `DOCKER_EXTENSIONS_NO_REGISTRATION` - basically, we can simply check if the extension has an `extension.json`  at the root and if it does not - use `require_once` instead of `wfLoadExtension`, something like that:

```php
foreach ( $dockerLoadExtensions as $extension ) {
	$jsonFile = "$wgExtensionDirectory/$extension/extension.json";
	if ( file_exists( $jsonFile ) ) {
		wfLoadExtension( $extension );
	} else {
		require_once "$wgExtensionDirectory/$extension/$extension.php";
	}
}
```

Reply:
I think file_exists() is too slow and we should not use it in this way. it can reduce the performance dramatically on heavy load servers

Reply:
is it really so? it looks like `file_exists` has a built-in cache [2] and according to this test [1] it took around 2.8 sec to check 1 000 000 files which probably could be roughly extrapolated to a `0.0000028` of a second per file, plus we have a bunch of `file_exists` calls across the core [3] I doubt this could make any noticeable difference

1. https://www.php.net/manual/en/function.file-exists.php#114011
2. https://www.php.net/manual/en/function.file-exists.php
3. https://github.com/wikimedia/mediawiki/search?p=5&q=file_exists

Reply:
makes sense, I measured this (only part of the code related to the extensions load)
on my computer (core i7 and SSD) it takes + 0.0002 sec (0.00126 vs 0.00147)
and on test.wikiteq.com it takes + 0,001 sec (0.00129 vs 0.00237)
not very much but it seems it does not cache it between sessions. I think it should be cached on OS side, but anyway IO operations is usually a bottleneck when we say about huge loaded systems.
are you sure still that we should use `file_exists` there?

Reply:
Yep, I think I'd vote to sacrifice this little optimization in favor of reducing cognitive complexity, especially considering there is a new settings approach where we could have an external settings file that needs to load the extensions by itself [1]

Yep, I think you're right that there is no caching between sessions [2], this is probably due to the nature of the PHP lifecycle

Speaking of hi-loads, I believe we'll hit bottlenecks in other places first before we'll reach the `file_exists` problem :) In the end, we could work around this by caching `file_exists` results by hand using CACHE_ACCEL or in-memory cache

1. https://github.com/WikiTeq/docker-test.wikiteq.com/blob/master/1.35.2/_settings/LocalSettings.php#L46
2. https://www.php.net/manual/en/function.clearstatcache.php

Reply:
Thank you!